op-program/client/driver/driver.go

```diff
@@ -72,11 +126,11 @@ func NewDriver(logger log.Logger, cfg *rollup.Config, l1Source derive.L1Fetcher,
 // Returns io.EOF if the derivation completed successfully
 // Returns a non-EOF error if the derivation failed
 func (d *Driver) Step(ctx context.Context) error {
-	if err := d.pipeline.Step(ctx); errors.Is(err, io.EOF) {
-		d.logger.Info("Derivation complete: reached L1 head", "head", d.engine.SafeL2Head())
+	if err := d.deriver.SyncStep(ctx); errors.Is(err, io.EOF) {
+		d.logger.Info("Derivation complete: reached L1 head", "head", d.deriver.SafeL2Head())
 		return io.EOF
 	} else if errors.Is(err, derive.NotEnoughData) {
-		head := d.engine.SafeL2Head()
+		head := d.deriver.SafeL2Head()
 		if head.Number >= d.targetBlockNum {
 			d.logger.Info("Derivation complete: reached L2 block", "head", head)
```

Comment:
This is why we're not stopping when the target block is reached anymore.  We used to get a `NotEnoughData` error each time the safe head advanced but no longer do. That is a rather odd assumption so we could do this check if there is no error returned as well.  On the mainnet genesis test (stopping after the first block import) we get no error 476 times vs getting NotEnoughData 3 times on the old code so we do wind up running this check a lot more often. However I think that's fine as we're just reading a field and doing a number comparison.

Reply:
I am trying to get rid of `NotEnoughData`, since it was specific to the EngineQueue. Good find, I'll try and fix the op-program driver.

Reply:
I'm fine with getting rid of it - it is a slightly weird error to return, particularly because it doesn't mean no progress was made. Just calling Step again without providing any new information would let it continue which is pretty weird. `io.EOF` is basically what I'd expect `NotEnoughData` to actually mean - I've derived everything I can with the data available, you need to provide more before you bother calling `Step` again.